File: src/main/java/br/com/simbasoft/renderer/Camera.java
package br.com.simbasoft.renderer;

public class Camera {
    private Vector3 position;
    private Vector3 direction;
    private Vector3 forwardVelocity;
    private double yaw;
    private double pitch;

    public Camera(Vector3 position, Vector3 direction) {
        this.position = position;
        this.direction = direction;
        this.forwardVelocity = new Vector3(0, 0, 0);
        this.yaw = 0;
        this.pitch = 0;
    }

    public Vector3 getLookAtTarget() {
        Vector3 target = new Vector3(0, 0, 1);

        Matrix4 cameraYawRotation = Matrix4.makeRotationY(yaw);
        Matrix4 cameraPitchRotation = Matrix4.makeRotationX(pitch);

        Matrix4 cameraRotation = Matrix4.makeIdentity();
        cameraRotation = Matrix4.multiply(cameraYawRotation, cameraRotation);
        cameraRotation = Matrix4.multiply(cameraPitchRotation, cameraRotation);

        Vector4 cameraDirection = Matrix4.multiplyVector4(cameraRotation, Vector4.fromVector3(target));
        direction = Vector3.fromVector4(cameraDirection);

        target = Vector3.add(position, direction);

        return target;
    }

    public void updateCameraPosition(Vector3 position) {
        this.position = position;
    }

    public void updateCameraForwardVelocity(Vector3 forwardVelocity) {
        this.forwardVelocity = forwardVelocity;
    }

    public void rotateCameraYaw(double angle) {
        this.yaw += angle;
    }

    public void rotateCameraPitch(double angle) {
        this.pitch += angle;
    }

    public Vector3 getPosition() {
        return position;
    }

    public Vector3 getDirection() {
        return direction;
    }

    public Vector3 getForwardVelocity() {
        return forwardVelocity;
    }
}

File: c/src/camera.c
#include "camera.h"
#include "matrix.h"

static camera_t camera;

void init_camera(vec3_t position, vec3_t direction) {
    camera.position = position;
    camera.direction = direction;
    camera.forward_velocity = vec3_new(0, 0, 0);
    camera.yaw = 0.0;
    camera.pitch = 0.0;
}

void update_camera_position(vec3_t new_position) {
    camera.position = new_position;
}

void update_camera_direction(vec3_t new_direction) {
    camera.direction = new_direction;
}

void update_camera_forward_velocity(vec3_t new_forward_velocity) {
    camera.forward_velocity = new_forward_velocity;
}

void rotate_camera_yaw(float angle) {
    camera.yaw += angle;
}

void rotate_camera_pitch(float angle) {
    camera.pitch += angle;
}

vec3_t get_camera_position(void) {
    return camera.position;
}

vec3_t get_camera_direction(void) {
    return camera.direction;
}

vec3_t get_camera_forward_velocity(void) {
    return camera.forward_velocity;
}

float get_camera_yaw(void) {
    return camera.yaw;
}

float get_camera_pitch(void) {
    return camera.pitch;
}

vec3_t get_camera_lookat_target(void) {
    vec3_t target = { 0, 0, 1 };

    mat4_t camera_yaw_rotation = mat4_make_rotation_y(camera.yaw);
    mat4_t camera_pitch_rotation = mat4_make_rotation_x(camera.pitch);

    mat4_t camera_rotation = mat4_identity();
    camera_rotation = mat4_mul_mat4(camera_yaw_rotation, camera_rotation);
    camera_rotation = mat4_mul_mat4(camera_pitch_rotation, camera_rotation);

    vec4_t camera_direction = mat4_mul_vec4(camera_rotation, vec4_from_vec3(target));
    camera.direction = vec3_from_vec4(camera_direction);

    target = vec3_add(camera.position, camera.direction);

    return target;
}